release-25.3: sql: add support for generating split points from table statistics

fqazi opened the pull request:
Backport 1/1 commits from #150238 on behalf of @fqazi.

----

Previously, we only injected split points before a backfill began, either at the start of the index span or by copying the splits from an index with matching key columns. This meant that split points were primarily added only when creating a duplicate index or adding a new column.

The bulk adder also has logic to create splits during `CREATE INDEX`, but its sampling method can lead to an insufficient number or poor placement of split points if the data samples from different nodes overlap.

To address this, this patch introduces the ability to create split points from table statistics, when available. This improves the split and scatter behavior for new indexes, especially those on columns that have existing statistics or a limited number of unique values.

Fixes: #148288
Release note (bug fix): Improved split and scatter behavior for CREATE INDEX when statistics are available for key columns. This can be enabled by setting the cluster setting: `schemachanger.backfiller.split_with_stats.enabled`

----

Release justification: this functionality is disabled by default and needs a cluster setting to explicitly enable.

fqazi committed:
sql: add support for generating split points from table statistics

Previously, we only injected split points before a backfill began,
either at the start of the index span or by copying the splits from an
index with matching key columns. This meant that split points were
primarily added only when creating a duplicate index or adding a new
column.

The bulk adder also has logic to create splits during `CREATE
INDEX`, but its sampling method can lead to an insufficient number or
poor placement of split points if the data samples from different nodes
overlap.

To address this, this patch introduces the ability to create
split points from table statistics, when available. This improves the
split and scatter behavior for new indexes, especially those on columns
that have existing statistics or a limited number of unique values.

Fixes: #148288
Release note (bug fix): Improved split and scatter behavior for CREATE
INDEX when statistics are available for key columns. This can be enabled
by setting the cluster setting:
`schemachanger.backfiller.split_with_stats.enabled`

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151838)
<!-- Reviewable:end -->

mw5h committed:
logic test: deflake distsql_stats

A recent change introduced by #150238 caused this test to sometimes
choose a local plan distribution when using the 5node logic test
configuration. This test is really more about the estimated row counts,
so constrain the the EXPLAIN statements to just look at row count
estimates instead of the entire plan.

Fixes: #151839
Release note: None
